homeassistant/components/input_datetime.py

```diff
@@ -0,0 +1,214 @@
+"""
+Component to offer a way to select a date and / or a time.
+
+For more details about this component, please refer to the documentation
+at https://home-assistant.io/components/input_datetime/
+"""
+import asyncio
+import logging
+import datetime
+
+import voluptuous as vol
+
+from homeassistant.const import ATTR_ENTITY_ID, CONF_ICON, CONF_NAME
+from homeassistant.loader import bind_hass
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.restore_state import async_get_last_state
+from homeassistant.util import dt as dt_util
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'input_datetime'
+ENTITY_ID_FORMAT = DOMAIN + '.{}'
+
+CONF_HAS_DATE = 'has_date'
+CONF_HAS_TIME = 'has_time'
+
+ATTR_DATE = 'date'
+ATTR_TIME = 'time'
+
+SERVICE_SET_DATETIME = 'set_datetime'
+
+SERVICE_SET_DATETIME_SCHEMA = vol.Schema({
+    vol.Required(ATTR_ENTITY_ID): cv.entity_ids,
+    vol.Optional(ATTR_DATE): cv.date,
+    vol.Optional(ATTR_TIME): cv.time,
+})
+
+
+def _cv_input_datetime(cfg):
+    """Configure validation helper for input datetime (voluptuous)."""
+    has_date = cfg[CONF_HAS_DATE]
+    has_time = cfg[CONF_HAS_TIME]
+
+    if not has_date and not has_time:
+        raise vol.Invalid("Input Datetime must have at least date or time!")
+    return cfg
```

Comment:
use vol.Exclusive (I'm not 100% sure) but the already exists a validate options to handle this case

Reply:
According to https://github.com/alecthomas/voluptuous/issues/126 , this functionality does not currently exist. vol.Exclusive() is 'XOR', we need 'OR', basically. The solution suggested in the issue is to have two schemata: one for validating the keys, one for validating the values. I could probably adapt this to something like this (with different names / constants): 
```
has_date_or_time_schema = Schema({
    Required(
        Any('has_time', 'has_date'),
        msg="Must have at least one of time or date!"): True
}, extra=ALLOW_EXTRA)

config_values_schema = Schema({
    Optional('has_time'): bool,
    Optional('has_date'): bool,
    … (all the rest) …
})

config_schema = All(has_date_or_time_schema, config_values_schema)
```
Would you consider that preferable? Or did I miss some better way?

Reply:
Maybe use https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/config_validation.py#L45?

Reply:
Ah, now that I try to use `has_at_least_one_key`: What I actually need to check is not that at least one of the two keys is present, but that at least one of the two is present *and* set to True. Otherwise, this allows configuring an input_datetime that carries neither time nor date.

I think the above example achieves that, the current code in the PR also does, but with `has_at_least_one_key` that is not possible. A third possibility that I see (additional to what's currently there and what I outlined above):

- Add an `as_at_least_one_key_value` to config_validation.py which takes a list of (key, value) pairs and validates that at least one of those is present in the dict.
- Use that with `[(CONF_HAS_TIME, True), (CONF_HAS_DATE, True)]` in the schema

Would that be an acceptable solution?